api/api/catalogue.go

```diff
@@ -21,8 +21,39 @@ type cubeResponse struct {
 	Sas              string `json:"sas,omitempty"    example:"<sas>"`
 }
 
+
+// Definition of marshalling and unmarshalling needs local type
+type Duration time.Duration
+
+func (duration Duration) MarshalJSON() ([]byte, error) {
+	return json.Marshal(time.Duration(duration).String())
+}
+
+/* We forbid float64 as duration type here as the interacting service should
+ * specify a duration with unit via a string.
+ */
+func (duration *Duration) UnmarshalJSON(b []byte) error {
+	var unmarshalledJson interface{}
+	if err := json.Unmarshal(b, &unmarshalledJson); err != nil {
+		return err
+	}
+
+	switch value := unmarshalledJson.(type) {
+	case string:
+		tmp, err := time.ParseDuration(value)
+		if err != nil {
+			return err
+		}
+		*duration = Duration(tmp)
+	default:
+		return fmt.Errorf("invalid duration: %#v", unmarshalledJson)
```

Comment:
Should we be more helpful to the user?
If I would get "invalid duration" I would be quite confused.

Default json error messages may not be extremely clueless-user-friendly, but they hint that the problem lies in type.
`{"error":"json: cannot unmarshal string into Go struct field SliceRequest.lineno of type int"}`

Could we also state something about expecting a string here?

Reply:
At the moment the error message is consistent with the error message `time` creates if one tries to create a `time.Duration` with invalid input and there is no other helpful error message to show. See, for example, the last error I get when trying to parse the string `asdf` in the opening comment of this PR.

Also, I have not found to actually trigger this error message. I only see error message that are created by the `time` package when I try to parse the value via the line
```
tmp, err := time.ParseDuration(value)
```

Nevertheless, better error messages definitely make sense. I am unsure what is the best way to phrase the message since I do not know how specific it should be to SAS token validity durations since the type name (`Duration`) at the moment is very general. I would personally like something in the following way
```suggestion
		return fmt.Errorf("catalogue: cannot unmarshal duration of SAS token %#v into duration type that expects string (digits+duration unit)\nExample: 60m, unmarshalledJson)
```
as it also gives and example to the correct format. In that case I would also be in favor to rename our special duration type to something like `sasTokenValidityDuration` to make clear that our duration type has a very specific use case.

In case we rename the duration type, I would also be happy to check for a valid duration length, e.g., between 1ms and 24h in the `UnmarshalJSON` method.

Reply:
Then I will rename the type.

I also agree with your remarks regarding the error message so I will change it.

Regarding not being able to trigger messages: I meant that I have not triggered the error message due to wrong type (yet) in my testing. When I test the code via a simple Python script, I always send strings in the request. That means all errors that I observe are actual conversion errors:

```
[GIN] 2022/11/23 - 12:03:14 | 400 |     214.166µs |             ::1 | POST     "/catalogue"
Error #01: time: missing unit in duration "60"
[GIN] 2022/11/23 - 12:03:14 | 400 |     197.875µs |             ::1 | POST     "/catalogue"
Error #01: time: invalid duration "asdf"
[GIN] 2022/11/23 - 12:03:14 | 400 |       152.5µs |             ::1 | POST     "/catalogue"
Error #01: time: missing unit in duration "60.0"
[GIN] 2022/11/23 - 12:03:14 | 400 |     131.958µs |             ::1 | POST     "/catalogue"
Error #01: time: unknown unit "kW" in duration "100kW"
```
 
Sorry for being unclear here. I am happy with the error messages here at the moment, but maybe it makes sense to change that in the near future. I could see that it makes life easier when/(if?) we expose the errors to the user to point directly to the SAS duration?